hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataTableValidator.java

```diff
@@ -500,19 +578,179 @@ public int compare(FileSlice o1, FileSlice o2) {
+        List<Pair<String, String>> partitionFileNameList = baseFileNameList.stream()
+            .map(filename -> Pair.of(partitionPath, filename)).collect(Collectors.toList());
+        return allColumnNameList.stream()
+            .flatMap(columnName ->
+                tableMetadata.getColumnStats(partitionFileNameList, columnName).values().stream()
+                    .map(stats -> new HoodieColumnRangeMetadata<>(
+                        stats.getFileName(),
+                        columnName,
+                        stats.getMinValue(),
+                        stats.getMaxValue(),
+                        stats.getNullCount(),
+                        stats.getValueCount(),
+                        stats.getTotalSize(),
+                        stats.getTotalUncompressedSize()))
+                    .collect(Collectors.toList())
+                    .stream())
+            .sorted(new HoodieColumnRangeMetadataComparator())
+            .collect(Collectors.toList());
+      } else {
+        return baseFileNameList.stream().flatMap(filename ->
+                new ParquetUtils().readRangeFromParquetMetadata(
+                    metaClient.getHadoopConf(),
+                    new Path(new Path(metaClient.getBasePath(), partitionPath), filename),
+                    allColumnNameList).stream())
+            .map(rangeMetadata -> new HoodieColumnRangeMetadata<String>(
+                rangeMetadata.getFilePath(),
+                rangeMetadata.getColumnName(),
+                // Note: here we ignore the type in the validation,
+                // since column stats from metadata table store the min/max values as String
+                rangeMetadata.getMinValue().toString(),
+                rangeMetadata.getMaxValue().toString(),
+                rangeMetadata.getNullCount(),
+                rangeMetadata.getValueCount(),
+                rangeMetadata.getTotalSize(),
+                rangeMetadata.getTotalUncompressedSize()
+            ))
+            .sorted(new HoodieColumnRangeMetadataComparator())
+            .collect(Collectors.toList());
+      }
+    }
+
+    public List<BloomFilterData> getSortedBloomFilterList(
+        String partitionPath, List<String> baseFileNameList) {
+      if (enableMetadataTable) {
+        List<Pair<String, String>> partitionFileNameList = baseFileNameList.stream()
+            .map(filename -> Pair.of(partitionPath, filename)).collect(Collectors.toList());
+        return tableMetadata.getBloomFilters(partitionFileNameList).entrySet().stream()
+            .map(entry -> BloomFilterData.builder()
+                .setPartitionPath(entry.getKey().getKey())
+                .setFilename(entry.getKey().getValue())
+                .setBloomFilter(entry.getValue())
+                .build())
+            .sorted()
+            .collect(Collectors.toList());
+      } else {
+        return baseFileNameList.stream()
+            .map(filename -> readBloomFilterFromFile(partitionPath, filename))
```

Comment:
same as 
```
  @Override
  public Map<Pair<String, String>, ByteBuffer> getBloomFilters(final List<Pair<String, String>> partitionNameFileNameList)
      throws HoodieMetadataException {
    throw new HoodieMetadataException("Unsupported operation: getBloomFilters!");
  }
```

If we could unify the API, we don' need to check `if (enableMetadataTable)` here.

Reply:
That's a good point as well.  I would tackle that in a separate PR and refactor the bloom index altogether: https://issues.apache.org/jira/browse/HUDI-3533